### PR TITLE
Implement Parallel TimSort Algorithm 

### PR DESCRIPTION
**Pull request by Yudhishthira1406**
Signed-off-by: Yudhishthira1406 <divyen1406@gmail.com>

**Parallel TimSort Implementation**
Implemented parallel insertion sort for each block and parallel merge each segment at each level of merging

- [x] implementation and documentation
- [x] correctness test
- [x] performance test

The issue #15045

### CODE REVIEW

**Comment by Yudhishthira1406** [comment]
Hi, how to perform tests for correctness and performance

**Comment by cassella** [comment]
Hi @Yudhishthira1406 ,

The Chapel test system is described in `doc/rst/developer/bestPractices/TestSystem.rst`.

The existing Sort correctness tests are in `test/library/packages/Sort/correctness`.  General correctness testing is in `correctness.chpl` there.  If there are any ways TimSort should be tested that aren't covered by that file, you can add other tests to that directory.

There are performance tests in `test/library/packages/Sort/performance`.  There's a `performance.chpl` there that tests the various sort types.  Note that there are supporting files that handle the metadata of performance tests, which will need to be modified as well, as described in the `TestSystem` document.

**Comment by Yudhishthira1406** [comment]
I have added performance and correctness test and the performance of timsort seems to be at par with default sorting algorithm QuickSort

File: test/library/packages/Sort/performance/performance.good [hunk 1]
```diff
@@ -6,3 +6,5 @@ insertionSort (seconds): x.xxx
 binaryInsertionSort (seconds): x.xxx
 mergeSort (seconds): x.xxx
 selectionSort (seconds): x.xxx
+timSort (seconds): x.xxx
+
```
Comment (ben-albrecht) [inline test/library/packages/Sort/performance/performance.good:10]: This added newline is causing a failure in testing

File: modules/packages/Sort.chpl [hunk 2]
```diff
@@ -796,6 +796,111 @@ module BinaryInsertionSort {
   }
 }
 
+pragma "no doc" 
+module TimSort {
+  private use Sort;
+
+    /*
+    Sort the 1D array `Data` using a parallel tim sort algorithm.
```
Comment (ben-albrecht) [inline modules/packages/Sort.chpl:804]: I think it's just 1 word:

```suggestion
    Sort the 1D array `Data` using a parallel timsort algorithm.
```

at least according to the original impl: https://github.com/python/cpython/blob/master/Objects/listsort.txt

**Review by ben-albrecht — COMMENTED** [review]
@Yudhishthira1406 - tests, docs, and overall structure of code looks great. I'll take a closer look at the implementation next. Thanks for the contribution!

**Comment by ben-albrecht** [comment]
Pasting some performance data gathered with default `perfexecopts` (sorting 16 MB of data) on my laptop for reference:

```
==> binaryInsertionSort.dat <==
5.2e-05

==> bubbleSort.dat <==
0.000255

==> heapSort.dat <==
0.355392

==> insertionSort.dat <==
5.7e-05

==> mergeSort.dat <==
0.130609

==> quickSort.dat <==
0.179038

==> radixsortMSB.dat <==
25.5153

==> selectionSort.dat <==
7.5e-05

==> timSort.dat <==
0.091562
```

(man, I wish these tests had a mode that spit out a markdown-flavored table)

File: test/library/packages/Sort/performance/performance.good (hunk 1 above)
Comment (Yudhishthira1406) [reply test/library/packages/Sort/performance/performance.good]: Thank you for review
UPD: fixed

File: modules/packages/Sort.chpl (hunk 2 above)
Comment (Yudhishthira1406) [reply modules/packages/Sort.chpl]: UPD:Fixed

**Comment by Yudhishthira1406** [comment]
Let me know if there is anything else
Thanks

**Comment by Yudhishthira1406** [comment]
Thanks @ben-albrecht for the review
I have made the suggested changes, if there is anything else let me know.

File: modules/packages/Sort.chpl [hunk 3]
```diff
@@ -796,6 +796,111 @@ module BinaryInsertionSort {
...
+    chpl_check_comparator(comparator, eltType);
+
+    if Dom.rank != 1 {
+      compilerError("timSort() requires 1-D array");
+    }
+
+    _TimSort(Data, Dom.alignedLow, Dom.alignedHigh, SIZE, comparator);
+  }
+
+  private proc _TimSort(Data: [?Dom], lo:int, hi:int, SIZE=16, comparator:?rec=defaultComparator) {
```
Comment (ben-albrecht) [inline modules/packages/Sort.chpl:825]: Let's rename `SIZE` to something more descriptive, like `blockSize`.

File: modules/packages/Sort.chpl [hunk 4]
```diff
@@ -796,6 +796,111 @@ module BinaryInsertionSort {
...
+    forall i in 0..#chunks {
+      InsertionSort.insertionSort(Data, comparator = comparator, lo + (i * SIZE) * stride, min(hi, lo + ((i + 1) * SIZE * stride) - stride));
+    }
+
+    /* apply merge operations on each block
+    *as the merges at a level are independent of each other
+    *they can be applied parallely 
+    */
+
+    var num_size = SIZE;
```
Comment (ben-albrecht) [inline modules/packages/Sort.chpl:844]: Chapel typically uses `camelCase` over `snake_case`. Let's call this `numSize`

File: modules/packages/Sort.chpl [hunk 5]
```diff
@@ -796,6 +796,111 @@ module BinaryInsertionSort {
...
+
+        _Merge(Data, l, mid, r, comparator=comparator);
+      }
+
+      num_size = num_size * 2;
+
+    }
+  }
+
+  private proc _Merge(Dst: [?Dom] ?eltType, lo:int, mid:int, hi:int, comparator:?rec=defaultComparator) {
```
Comment (ben-albrecht) [inline modules/packages/Sort.chpl:860]: Is this any different from `private proc _Merge` in `MergeSort` module? If not, I think we should make the `MergeSort` one public (but add a `pragma "no doc" so it doesn't show up in public docs), and re-use that, similar to how we re-use insertion sort.

**Review by ben-albrecht — CHANGES_REQUESTED** [review]
Dove into the implementation a bit and have some comments. Thanks!

File: modules/packages/Sort.chpl (hunk 5 above)
Comment (Yudhishthira1406) [reply modules/packages/Sort.chpl:860]: I have made few changes from `private proc _Merge` in `MergeSort`. In MergeSort it alternates the storage of Segments in the original memory and the copied memory to merge, but for parallelism, it was required to create a copy inside the `private proc _Merge` and store the results back into the original memory

Comment (ben-albrecht) [reply modules/packages/Sort.chpl:860]: OK, then let's add a doc-string above this line to describe the diff between this `_Merge` and `MergeSort`'s version.

File: modules/packages/Sort.chpl [hunk 6]
```diff
@@ -796,6 +796,111 @@ module BinaryInsertionSort {
   }
 }
 
+pragma "no doc" 
+module TimSort {
+  private use Sort;
+
+    /*
+    Sort the 1D array `Data` using a parallel timSort algorithm.
+
```
[the remark this answers is not in the document]
Comment (Yudhishthira1406) [reply modules/packages/Sort.chpl:805]: Does this seem good
Screenshot from 2021-01-21 22-51-02

[the remark this answers is not in the document]
Comment (ben-albrecht) [reply modules/packages/Sort.chpl:805]: Yes, thanks.

**Comment by Yudhishthira1406** [comment]
@ben-albrecht
I made the suggested changes

File: modules/packages/Sort.chpl (hunk 5 above)
Comment (Yudhishthira1406) [reply modules/packages/Sort.chpl:862]:
```
    The `private proc _Merge` is disfferent from the `private proc _Merge` in `MergeSort`.
    In MergeSort it alternates the storage of segments in the original memory and the copied memory. 
    In the below defined `private proc _Merge`, it creates a copy of the segments to be merged and 
    store the results back into the original memory
```
Is this ok?

Comment (ben-albrecht) [reply modules/packages/Sort.chpl:862]: Looks good, see some suggested edits below:

```
   This TimSort._Merge() differs from MergeSort._Merge() in the following way:
   MergeSort._Merge() alternates the storage of segments in the original memory and the copied memory. 
   TimSort._Merge() creates a copy of the segments to be merged and 
   stores the results back into the original memory.
```

Comment (Yudhishthira1406) [reply modules/packages/Sort.chpl:868]: Done

File: modules/packages/Sort.chpl [hunk 7]
```diff
@@ -796,6 +796,119 @@ module BinaryInsertionSort {
...
+    }
+
+    /* apply merge operations on each block
+    *as the merges at a level are independent of each other
+    *they can be applied parallely 
+    */
+
+    var numSize = blockSize;
+    while(numSize < size) {
+      forall i in 0..(size - 1) by 2 * numSize {
```
Comment (ben-albrecht) [inline modules/packages/Sort.chpl:848]: This syntax is a little cleaner:

```suggestion
      forall i in 0..<size  by 2 * numSize {
```

File: modules/packages/Sort.chpl [hunk 8]
```diff
@@ -796,6 +796,119 @@ module BinaryInsertionSort {
...
+
+        var l = lo + i * stride;
```
Comment (ben-albrecht) [inline modules/packages/Sort.chpl:850]: Good practice to make vars that you don't expect to change a `const`, e.g.

```suggestion
        const l = lo + i * stride;
```

**Review by ben-albrecht — COMMENTED** [review]
@Yudhishthira1406 - looks good to me. Made a couple more minor code quality suggestions.

Let me know when you're done making changes and I'll run testing before merging.

Thanks!

**Comment by Yudhishthira1406** [comment]
@ben-albrecht 
I have made the suggested changes to code quality.
Thanks

**Review by ben-albrecht — APPROVED** [review]
LGTM.

Double checked that testing passes:

```
> start_test test/library/packages/Sort
...
[Test Summary - 210122.095325]
[Summary: #Successes = 50 | #Failures = 0 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]
```


Thank you for contributing @Yudhishthira1406!

**Comment by ben-albrecht** [comment]
@Yudhishthira1406 - I will plan to merge on Monday in case weekend failures. Thanks!

**Comment by damianmoz** [comment]
Thanks for the development work.

Has any investigation been done on the parallelism achieved/achievable with this implementation.  Is there some balancing guidelines for the choice of **blocksize** given the number of processors available to the user and the amount of cache within those processors? What balance is there between **blocksize** and the depth of the **merge** tree? I assume that the absence of any use of **galloping** is driven by simplicity?

I have to admit that when I went looking for non-Chapel example code on the net, I was disappointed at its quality. Reference code written in Python somehow lost an element of the input. Another example in C using a binary insertion sort was not sorted correctly. 

I found the alternating use of square brackets and round parentheses for indexing of the arrays confusing. Just my 2c.

Also, are there tighter ways to write
```chapel
    while (a1 <= a1max) {
      Dst[i] = A1[a1];
      a1 += stride;
      i += stride;
    }
```

**Comment by ben-albrecht** [comment]
Hey @damianmoz, I'll let @Yudhishthira1406 answer if they tried tuning `blocksize` and number of tasks.  If not, that may be a low-hanging fruit for performance tuning. 

> I found the alternating use of square brackets and round parentheses for indexing of the arrays confusing.

I agree, and tend to always use `[]` for indexing.

> Also, are there tighter ways to write ...

If you calculated the ranges up front, you could do an array slice assignment:

```
Dst[i..imax by stride] = A1[a1..a1max by stride];
```

**Comment by damianmoz** [comment]
Does Chapel protect its array assignments? Those assignments need to be
```Chapel
if a1 <= a1max then
{
   Dst[i..imax by stride] = A1[a1..a1max by stride]
}
if a2 <= a2max then
{
   Dst[i..imax by stride] = A2[a2..a2max by stride]
}
```
Are those **if** statements redundant?

Also, the sort module is riddled with an argument specification
```Chapel
comparator=comparator
```
even when that argument is in the location where you should be able to say simply
```Chapel
comparator
```
That seems like it is being overly complicated or am I missing something?

**Comment by ben-albrecht** [comment]
> Are those if statements redundant?

That's a good point. `a1..a1max` would be an empty range and iterate over no elements if `a1 > a1max`, so yes -- that loop could be simplified further.

> That seems like it is being overly complicated or am I missing something?

I think this comes down to personal preference. Because the comparators often have "comparator" in their name, I'd agree with you in that the `comparator=` argument name could be dropped in the calls.

**Pull request merged**